src/internal/collection/postgres_listener.go

```diff
@@ -0,0 +1,395 @@
+package collection
+
+import (
+	"context"
+	"encoding/base64"
+	"math"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/lib/pq"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+)
+
+const (
+	minReconnectInterval = time.Second * 1
+	maxReconnectInterval = time.Second * 30
+	channelBufferSize    = 1000
+)
+
+type postgresWatcher struct {
+	listener   *PostgresListener
+	c          chan *watch.Event
+	buf        chan *postgresEvent // buffer for messages before the initial table list is complete
+	done       chan struct{}       // closed when the watcher is closed to interrupt selects
+	template   proto.Message
+	sqlChannel string
+
+	// Filtering variables:
+	opts      watch.WatchOptions // may filter by the operation type (put or delete)
+	index     *string            // only set if the watch filters by an index (for dealing with hash collisions)
+	value     *string            // only set if 'index' is set
+	startTime *time.Time         // only set once the initial list has completed and we know the start time for the watch
+}
+
+func newPostgresWatch(
+	listener *PostgresListener,
+	sqlChannel string,
+	template proto.Message,
+	index *string,
+	value *string,
+	opts watch.WatchOptions,
+) *postgresWatcher {
+	return &postgresWatcher{
+		listener:   listener,
+		c:          make(chan *watch.Event),
+		buf:        make(chan *postgresEvent, channelBufferSize),
+		done:       make(chan struct{}),
+		template:   template,
+		sqlChannel: sqlChannel,
+		opts:       opts,
+		index:      index,
+		value:      value,
+	}
+}
+
+func (pw *postgresWatcher) Watch() <-chan *watch.Event {
+	return pw.c
+}
+
+func (pw *postgresWatcher) Close() {
+	// Close the 'done' channel to interrupt any waiting writes - because of this,
```

Comment:
This comment needs to be visible to users either on the interface or above the method.  Alternatively make it safe using sync.Once.